distsql: a couple of fixes

yuzefovich opened the pull request:
I extracted two commits that we want to merge from #38952.

First commit clears vectorized memory account on setup error.
Otherwise there could be cases where memory reservations could live on
in the monitor since Cleanup would never be called on a flow.

Second commit avoids double freeing processor specs.
Distributed vectorized fallback logic introduced recursive calls to
setupFlows in which Releases of processor specs would be deferred. This
could result in double Releases of processor specs, in between which a
goroutine could have grabbed a spec from the pool, written to it, and
then found that the spec had been zeroed out.

Fixes: #38651.

Commit:
exec: clear vectorized memory account on setup error

Otherwise there could be cases where memory reservations could live on
in the monitor since Cleanup would never be called on a flow.

Release note: None

yuzefovich committed:
distsql: avoid double freeing processor specs

Distributed vectorized fallback logic introduced recursive calls to
setupFlows in which Releases of processor specs would be deferred. This
could result in double Releases of processor specs, in between which a
goroutine could have grabbed a spec from the pool, written to it, and
then found that the spec had been zeroed out.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39061)
<!-- Reviewable:end -->

asubiotto reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I think we probably want to remove any recursion here, right?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39061#-:-LkX00228A4lzqlNu7io:b-kkqzy6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
I'm not sure about removing the recursion part. I'm still actually leaning towards merging your full Fallback fixes PR because today I observed a case when our up-front check for vectorized support succeeded, but the execution failed during runtime. https://github.com/cockroachdb/cockroach/issues/38919#issuecomment-514325820

If we remove the recursion (and I assume you're suggesting also removing the fallback to DistSQL when a vectorized runtime execution error is observed?), we will have to err out on such cases.

jordanlewis commented:
Can we make that runtime failure case fail up front instead? It would be better to not have to have the distributed fallback stuff - IMO, if we fail once execution starts, that's it - time to give up and propagate the error all the way back to the user.

yuzefovich committed:
WIP on removing fallback logic

We now do the upfront check of vectorization support for a flow
almost equivalently to an actual flow setup. This should catch
possible runtime execution errors (like unsupported schema for
the outboxes).

Release note: None

yuzefovich commented:
I've just pushed another commit to this PR that is prototyping how we could remove the fallback logic and attempt to catch runtime execution errors (like the one I mentioned above). The idea is to simulate as close as possible what we do in `setupVectorized` to what we check in `SupportsVectorized`. Please let me know what you think.

jordanlewis reviewed (commented):
The third commit would obviate the first, right?

I think this is generally a great idea, but the implementation feels unideal. It would be better if we could decouple the creation of the operators (the parts that can return unsupported errors) from actually starting them (the part that we don't care to run, that you've prevented from running by passing in a nil Flow). That might be a larger refactor though. Note how it parallels what we've done elsewhere - we don't *start* any resources inside of constructors - we wait until Init time (or later).

If there's no path to fixing this right now, at least add some comments as to why we're passing nil flows, and what nil flows mean to each of the functions.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39061#-:-LkZ57Io7bDSRpMJEurl:b-3n5c21)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/distsql_running.go, line 223 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39061#-LkZ66IM4RvnsFoybxvc:-LkZ66IM4RvnsFoybxvd:b5qei0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2b7cb84feaf20028c4b3b05c6fde10b1dd1567/pkg/sql/distsql_running.go#L223)):*
> ```Go
> 			// Error vectorizing remote flows, and since we choose to not fall back
> 			// to DistSQL, we will return an error.
> 			log.VEventf(ctx, 1, "vectorization of flows with an error %+v", firstErr)
> ```

missing a verb (probably want "failed")
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 951 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39061#-LkZ6Zb076Ad5j-e6mzv:-LkZ6Zb076Ad5j-e6mzw:bw5dtvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2b7cb84feaf20028c4b3b05c6fde10b1dd1567/pkg/sql/distsqlrun/column_exec_setup.go#L951)):*
> ```Go
> // an error is returned if this does not happen (i.e. no router outputs are
> // remote).
> func setupVectorizedRouter(
> ```

You don't strictly have to change the receivers on all of these - it's okay to call a method on a nil receiver.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Actually the third commit would remove the need for both the first and the second.

I will look into doing the refactor now and will open up a separate PR, so I'm closing this one.
